src/lib.rs

```diff
@@ -193,19 +193,31 @@ impl LeakyBuckets {
 
         Ok(async move {
             let mut futures = FuturesUnordered::new();
+            let mut new_task_ended = false;
 
-            loop {
+            while !futures.is_empty() || !new_task_ended {
                 tokio::select! {
-                    new_task = new_task_rx.recv() => {
-                        let NewTask { inner, task_rx } = new_task.expect("new task queue ended");
+                    new_task = new_task_rx.recv(), if !new_task_ended => {
```

Comment:
`new_task_ended` seems to be always `false` here.

Reply:
It shouldn't be? It's set to `true` just below here once `new_task_rx` ends.